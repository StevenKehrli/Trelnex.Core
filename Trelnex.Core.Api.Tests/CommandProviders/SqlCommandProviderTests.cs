using Azure.Core;
using Azure.Identity;
using Microsoft.Data.SqlClient;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using Trelnex.Core.Api.CommandProviders;
using Trelnex.Core.Api.Serilog;
using Trelnex.Core.Data;
using Trelnex.Core.Data.Tests.CommandProviders;

namespace Trelnex.Core.Api.Tests.CommandProviders;

[Ignore("Requires a SQL server.")]
public class SqlCommandProviderTests : CommandProviderTests
{
    private readonly string _scope = "https://database.windows.net/.default";

    private TokenCredential _tokenCredential = null!;
    private string _connectionString = null!;
    private string _tableName = null!;

    [OneTimeSetUp]
    public void TestFixtureSetup()
    {
        // This method is called once prior to executing any of the tests in the fixture.

        // create the service collection
        var services = new ServiceCollection();

        // create the test configuration
        var configuration = new ConfigurationBuilder()
            .AddJsonFile("appsettings.json", optional: true, reloadOnChange: true)
            .AddJsonFile("appsettings.User.json", optional: true, reloadOnChange: true)
            .Build();

        var dataSource = configuration
            .GetSection("SqlCommandProviders:DataSource")
            .Value;

        var initialCatalog = configuration
            .GetSection("SqlCommandProviders:InitialCatalog")
            .Value;

        _tableName = configuration
            .GetSection("SqlCommandProviders:Tables:0:TableName")
            .Value!;

        var scsBuilder = new SqlConnectionStringBuilder()
        {
            DataSource = dataSource,
            InitialCatalog = initialCatalog,
            Encrypt = true,
        };

        _connectionString = scsBuilder.ConnectionString;

        // create the command provider
        _tokenCredential = new DefaultAzureCredential();

        var bootstrapLogger = services.AddSerilog(
            configuration,
            "Trelnex.Integration.Tests");

        services.AddSqlCommandProviders(
            configuration,
            bootstrapLogger,
            options => options.Add<ITestItem, TestItem>(
                typeName: "test-item",
                validator: TestItem.Validator,
                commandOperations: CommandOperations.All));

        var serviceProvider = services.BuildServiceProvider();

        // get the command provider
        _commandProvider = serviceProvider.GetRequiredService<ICommandProvider<ITestItem>>();
    }

    [TearDown]
    public void Cleanup()
    {
        // This method is called after each test has run.
        using var sqlConnection = new SqlConnection(_connectionString);

        var tokenRequestContext = new TokenRequestContext([ _scope ]);
        sqlConnection.AccessToken = _tokenCredential.GetToken(tokenRequestContext, default).Token;

        sqlConnection.Open();

        var cmdText = $"DELETE FROM [{_tableName}-events]; DELETE FROM [{_tableName}];";
        var sqlCommand = new SqlCommand(cmdText, sqlConnection);

        sqlCommand.ExecuteNonQuery();
    }
}
