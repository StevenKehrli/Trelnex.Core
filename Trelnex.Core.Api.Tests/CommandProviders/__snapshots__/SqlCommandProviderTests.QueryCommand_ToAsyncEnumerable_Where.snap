[
  {
    "Item": {
      "Id": "fc723d27-bb04-475a-8e77-01631518d0a1",
      "PartitionKey": "78b5102c-1cee-44e4-8f89-6b71cb8592b2",
      "TypeName": "test-item",
      "CreatedDate": "2025-01-04T03:34:10.537488Z",
      "UpdatedDate": "2025-01-04T03:34:10.537488Z",
      "DeletedDate": null,
      "IsDeleted": null,
      "ETag": "2994f805-da7f-43e7-8cd8-93410eeefe4f",
      "Message": "Message #1"
    }
  }
]
